File: app/src/main/java/com/udbstudents/tseapp/views/MainActivity.java
package com.udbstudents.tseapp.views;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.udbstudents.tseapp.R;
import com.udbstudents.tseapp.RetrofitService;
import com.udbstudents.tseapp.models.User;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {

    private final String baseurl = "https://172.09.09.9::8080/";
    private Retrofit retrofit = null;
    List<User>  userList = new ArrayList<>();

    private FloatingActionButton fav;
    private EditText mEditUser;
    private EditText mEditPassword;
    private Button btnLogin;
    private String userName = "";
    private String passwordText = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mEditUser = findViewById(R.id.edit_usuario);
        mEditPassword = findViewById(R.id.edit_pass);
        btnLogin = findViewById(R.id.btn_login);


        retrofit();
        retrofit = new Retrofit.Builder().baseUrl(baseurl)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        btnLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                userName = mEditUser.getText().toString().trim();
                passwordText = mEditPassword.getText().toString().trim();

                if (!userName.isEmpty() && !passwordText.isEmpty()){
                    signIn(userName, passwordText);
                }
                else{
                    Toast.makeText(MainActivity.this, "Los campos son necesarios", Toast.LENGTH_LONG).show();
                }
            }
        });

        fav = findViewById(R.id.fab);
        fav.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent myIntent = new Intent(MainActivity.this, SingUpActivity.class);
                startActivity(myIntent);
            }
        });


    }

    private void signIn(String userName, String passwordText) {
    }

    private void retrofit() {

        RetrofitService retrofitService = retrofit.create(RetrofitService.class);

        final Call<List<User>> listaUser = retrofitService.getUsers();

        listaUser.enqueue(new Callback<List<User>>() {
            @Override
            public void onResponse(Call<List<User>> call, Response<List<User>> response) {

                if (response.isSuccessful()){
                    userList = response.body();
                }
            }

            @Override
            public void onFailure(Call<List<User>> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/java/com/udbstudents/tseapp/views/MainActivity.kt
package com.udbstudents.tseapp.views

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.WindowManager
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.iid.FirebaseInstanceId
import com.udbstudents.tseapp.R
import com.udbstudents.tseapp.models.TokenUser
import com.udbstudents.tseapp.utils.FN
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*

class MainActivity : AppCompatActivity() {

    private lateinit var mAuth: FirebaseAuth
    private lateinit var mFirestore: FirebaseFirestore

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)

        mAuth = FirebaseAuth.getInstance()
        mFirestore = FirebaseFirestore.getInstance()

        btn_login.setOnClickListener {
            progressBarMain.visibility = View.VISIBLE
            btn_login.visibility = View.INVISIBLE
            if (edit_usuario.text.toString().trim().isNotEmpty() && edit_pass.text.toString().trim().isNotEmpty()) {
                logIn(edit_usuario.text.toString().trim(), edit_pass.text.toString().trim())
            } else {
                FN.alertProcesses(this, getString(R.string.empty_fields))
                progressBarMain.visibility = View.INVISIBLE
                btn_login.visibility = View.VISIBLE
            }
        }

        fab.setOnClickListener {
            startActivity(Intent(this, SingUpActivity::class.java))
            finish()
        }
    }

    private fun logIn(email: String, password: String) {
        mAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(this) {
            progressBarMain.visibility = View.INVISIBLE
            if (it.isSuccessful) {
                val user = mAuth.currentUser
                if (user != null) {
                    getTokenFromCloudMessagingService(user)
                } else {
                    updateUI(null)
                    FN.alertProcesses(this, getString(R.string.message_error_login))
                    progressBarMain.visibility = View.INVISIBLE
                    btn_login.visibility = View.VISIBLE
                }
            } else {
                updateUI(null)
                FN.alertProcesses(this, getString(R.string.message_error_login))
                progressBarMain.visibility = View.INVISIBLE
                btn_login.visibility = View.VISIBLE
            }
        }
    }

    private fun updateUI(user: FirebaseUser?) {
        if (user != null) {
            startActivity(Intent(this, PrincipalActivity::class.java))
            finish()
        } else {
            Log.i("Error Session", getString(R.string.warning_message))
        }
    }

    override fun onStart() {
        super.onStart()
        val currentUser = mAuth.currentUser
        if (currentUser != null) {
            updateUI(currentUser)
        } else {
            updateUI(null)
        }
    }

    private fun getTokenFromCloudMessagingService(userAuth: FirebaseUser?) {
        FirebaseInstanceId.getInstance().instanceId.addOnCompleteListener(OnCompleteListener { task ->
            if (!task.isSuccessful) {
                Log.w("token", "getInstanceId failed", task.exception)
                return@OnCompleteListener
            } else {
                val token = task.result?.token
                val user = userAuth?.uid
                val device = "Android"

                val tokenUser = TokenUser(token, device, user)
                FirebaseFirestore.getInstance().collection("TokenUsuario").whereEqualTo("deviceToken", token).get().addOnCompleteListener(this) {
                    if (it.isSuccessful) {
                        val tokenUserModel = it.result!!.toObjects(TokenUser::class.java)
                        if (tokenUserModel.size == 0) {
                            FirebaseFirestore.getInstance().collection("TokenUsuario").document().set(tokenUser)
                            updateUI(userAuth)
                        }
                    } else {
                        Log.i("token_exist", token!!)
                    }
                }.addOnFailureListener(this){
                    Log.e("ErrorTokenInserted", it.message!!)
                }
            }
        })
    }

}
